src/data/bucket/symbol_bucket.js

```diff
@@ -324,6 +325,14 @@ class SymbolBucket implements Bucket {
     }
 
     createArrays() {
+        const layout = this.layers[0].layout;
+        this.hasPaintOverrides = SymbolStyleLayer.hasPaintOverrides(layout);
+        if (this.hasPaintOverrides) {
+            for (const layer of this.layers) {
+                layer.setPaintOverrides();
```

Comment:
It looks like a SymbolStyleLayer can exist either with overrides applied or not. And since the state depends on this method call it is hard to tell elsewhere what the state of the style layer is. Your current implementation looks correct but I think someone less familiar with this could easily make changes that either result in `setPaintOverrides()` being called more or less than once, or before `recalculate()`. 

Would triggering it in `styleLayer.recalculate(...)` make sense?